### PR TITLE
Add Mentions, and generate Candidates based on Mentions

### PR DESCRIPTION
**Pull request by lukehsiao**
This also includes some improvements to tests and a bugfix for the Meta class.

### CODE REVIEW

File: CHANGELOG.rst [hunk 1]
```diff
@@ -1,26 +1,89 @@
-Version 0.3.0 (coming soon...)
-------------------------------
-
-.. warning::
-    This release is NOT backwards compatable with v0.2.3. Specifically we have
-    removed the ``get_parent()``, ``get_children()``, and
-    ``get_sentence_generator()`` functions from the classes inhereting from
-    Context in the data model.
-
+[Unreleased]
```
Comment (lukehsiao) [inline CHANGELOG.rst]: Made updates to the CHANGELOG to follow [Keep A Changelog](https://keepachangelog.com/en/1.0.0/) conventions

File: fonduer/candidates/candidates.py [hunk 2]
```diff
@@ -96,7 +21,7 @@ class CandidateExtractor(UDFRunner):
         one for each relation argument. Only tuples of Contexts for which each
         element is accepted by the corresponding Matcher will be returned as
         Candidates
-    :param candidate_filter: an optional function for filtering out candidates
+    :param candidate_throttler: an optional function for filtering out candidates
```
Comment (lukehsiao) [inline fonduer/candidates/candidates.py]: Renamed these to more closely match the keywords we use in the paper.

File: fonduer/candidates/models/mention.py [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+
+    def __gt__(self, other):
+        # Allow sorting by comparing the string representations of each
+        return self.__repr__() > other.__repr__()
+
+
+# This global dictionary contains all classes that have been declared in this
+# Python environment, so that mention_subclass() can return a class if it
+# already exists and is identical in specification to the requested class
+mention_subclasses = {}
```
Comment (SenWu) [inline fonduer/candidates/models/mention.py]: Why is this here? It should be somehwere high level?

File: fonduer/features/core_features.py [hunk 4]
```diff
@@ -9,15 +9,15 @@
...
+        args = tuple([arg.span for arg in candidate.get_contexts()])
         if not (isinstance(args[0], TemporarySpan)):
             raise ValueError(
                 "Accepts Span-type arguments, %s-type found." % type(candidate)
             )
 
         # Unary candidates
         if len(args) == 1:
-            span = args[0]
+            span = args[0].span
```
Comment (SenWu) [inline fonduer/features/core_features.py]: args is already a list of span?

File: fonduer/utils/data_model_utils/structural.py [hunk 5]
```diff
@@ -0,0 +1,198 @@
+#############################
+# Structural modality helpers
```
Comment (SenWu) [inline fonduer/utils/data_model_utils/structural.py]: Structural modality utilities? Better name

File: fonduer/utils/data_model_utils/structural.py [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+from lxml import etree
+from lxml.html import fromstring
+
+from fonduer.utils.data_model_utils.utils import _to_span
+
+
+def get_tag(mention):
+    """Return the HTML tag of the Mention.
+
+    If a candidate is passed in, only the tag of its first Mention is returned.
```
Comment (SenWu) [inline fonduer/utils/data_model_utils/structural.py]: candidate?

Comment (SenWu) [reply fonduer/utils/data_model_utils/structural.py]: I see. Should we return a list of tags? thoughts?

File: fonduer/utils/data_model_utils/tabular.py [hunk 7]
```diff
@@ -1,14 +1,14 @@
-############################
+##########################
 # Tabular modality helpers
```
Comment (SenWu) [inline fonduer/utils/data_model_utils/tabular.py]: utilities?

File: fonduer/utils/data_model_utils/textual.py [hunk 8]
```diff
@@ -1,18 +1,18 @@
-############################
+#########################
 # Visual modality helpers
```
Comment (SenWu) [inline fonduer/utils/data_model_utils/textual.py]: utilities?

File: tests/candidates/download_data.sh [hunk 9]
```diff
@@ -0,0 +1,16 @@
+echo "Downloading hardware tutorial data..."
```
Comment (SenWu) [inline tests/candidates/download_data.sh]: Can we use the same set of data so we only download once?

File: CHANGELOG.rst (hunk 1 above)
Comment (SenWu) [reply CHANGELOG.rst]: 👍

File: fonduer/candidates/candidates.py [hunk 10]
```diff
@@ -155,161 +81,97 @@ class CandidateExtractorUDF(UDF):
...
-        """
-        Generate OmniFigures from a Document by parsing all of its Figures.
-        """
-        if not isinstance(context, Document):
-            raise TypeError(
-                "Input Contexts to OmniFigures.apply() must be of type Document"
+        logger.info("Document: {}".format(context))
+        # Iterate over each candidate class
+        for i, candidate_class in enumerate(self.candidate_classes):
+            logger.info("  Relation: {}".format(candidate_class.__name__))
```
[the remark this answers is not in the document]
Comment (lukehsiao) [reply fonduer/candidates/candidates.py]: This just formats a little cleaner in the output (indented under the document)

File: fonduer/utils/data_model_utils/structural.py (hunk 6 above)
Comment (lukehsiao) [reply fonduer/utils/data_model_utils/structural.py]: I think it is ok as is.

File: fonduer/candidates/models/mention.py (hunk 3 above)
Comment (lukehsiao) [reply fonduer/candidates/models/mention.py]: Will move up in the file.

File: fonduer/candidates/candidates.py [hunk 11]
```diff
@@ -155,161 +81,97 @@ class CandidateExtractorUDF(UDF):
...
-        for figure in doc.figures:
-            if self.type is None or figure.url.lower().endswith(self.type):
-                yield TemporaryImage(figure)
+            for k, cand in enumerate(cands):
+
+                # Apply candidate_throttler if one was given.
+                # Accepts a tuple of Context objects (e.g., (Span, Span))
+                # (candidate_throttler returns whether or not proposed candidate
+                # passes throttling condition)
+                if self.candidate_throttlers[i]:
```
[the remark this answers is not in the document]
Comment (lukehsiao) [reply fonduer/candidates/candidates.py]: Currently tuple of spans. Will switch to tuple of Mentions

File: fonduer/candidates/candidates.py [hunk 12]
```diff
@@ -155,161 +81,97 @@ class CandidateExtractorUDF(UDF):
...
+
+                # TODO: Make this work for higher-order relations
+                if self.arities[i] == 2:
+                    ai, a = (cand[0][0], cand[0][1].span)
+                    bi, b = (cand[1][0], cand[1][1].span)
+
+                    # Check for self-joins, "nested" joins (joins from span to
+                    # its subspan), and flipped duplicate "symmetric" relations
+                    if not self.self_relations and a == b:
+                        logger.info("Skipping self-joined candidate {}".format(cand))
```
[the remark this answers is not in the document]
Comment (lukehsiao) [reply fonduer/candidates/candidates.py]: Switched to debug level logging so it won't show up unless logging is set to debug level.

File: fonduer/candidates/candidates.py [hunk 13]
```diff
@@ -112,37 +37,38 @@ class CandidateExtractor(UDFRunner):
...
-        logger.info("Clearing {}".format(self.candidate_class.__tablename__))
-        session.query(Candidate).filter(
-            Candidate.type == self.candidate_class.__tablename__
-        ).filter(Candidate.split == split).delete()
+        """Delete Candidates of each cass from given split the database."""
+        for candidate_class in self.candidate_classes:
+            logger.info("Clearing {}".format(candidate_class.__tablename__))
+            session.query(Candidate).filter(
+                Candidate.type == candidate_class.__tablename__
+            ).filter(Candidate.split == split).delete()
```
[the remark this answers is not in the document]
Comment (lukehsiao) [reply fonduer/candidates/candidates.py]: Will not affect Mention. Added new test case to show that.

File: Makefile [hunk 14]
```diff
@@ -19,9 +19,13 @@ check:
 	#   - E741: do not use variables named ‘l’, ‘O’, or ‘I’
 	#   - E203: whitespace before ‘:’
 	flake8 fonduer/ --count --max-line-length=127 --statistics --ignore=E731,W503,E741,E203
```
[the remark this answers is not in the document]
Comment (lukehsiao) [reply Makefile]: Going to try and bring these all down to max length of 88, which is the default max-line-length of `black`.
